### DISCUSSION TITLE
How to update an Alpine component when loaded by ajax

### DISCUSSION THREAD

**Discussion by jsantari** [alpinejs/alpine#857] [4. General]
I've been pulling my hair out for days on this. I have been trying dynamically load a component that uses an external data function and the html that goes with it using ajax using the htmx.org library. If all the code is in the initial page it works fine. When using the htmx.org library to load the js function code and html dynamically you always get a function not found error from Alpine on my function (named calc) as it apparently can not see it. If I move the js code into the main page and still load the html with htmx then it works but that defeats the purpose of trying to make it into a standalone component package. I've tried to define my function using window.calc = function() {...} but no luck with that either.  Is there no way to get this to work?

**Comment by BartChris** [comment]
Have you used
` <script type="text/javascript">`
inside your htmx-Template? The "text/javascript" is necessary for htmx to actually evalute the javascript code.

**Reply by jsantari** [reply]
I've tried that and it still doesn't work right. In my index.php file I have this htmx code to load my component based on the Alpine dropdown modal example.
```
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>AlpineJS/Tailwind Testing</title>
    <meta name="author" content="David Grzyb">
    <meta name="description" content="">

    <script src="js/htmx.js"></script>
    <script src="js/alpine.js"></script>

</head>

<body style="padding: 10px;">
  Test 1
   <div id="demo" hx-get="component.html" hx-trigger="load"></div>
</body>
</html>
```
My component.html has this Alpine code.
```

<div x-data="dropdown()">
    <button style="border-radius: 5px; border: 2px solid #0099FF; padding: 5px 10px;" x-on:click="open">Open</button>

    <div x-show="isOpen()" x-on:click.away="close">
        // Dropdown content
    </div>
</div>

<script type="text/javascript">
console.log('script load'); // added just to test

function dropdown() {
        return {
            show: false,
            open() { this.show = true },
            close() { this.show = false },
            isOpen() { return this.show === true },
        }
    }
</script>
```
The console log I put in just to see if it executed any of the script which is does. Htmx loads the component as I see the button and the modal content as it isn't initializing the function dropdown so the this.show value is not false and the modal content appears. However Alpine throws this error in the console.
```
alpine.js:1873 Uncaught ReferenceError: dropdown is not defined
    at eval (eval at saferEval (alpine.js:118), <anonymous>:3:36)
    at saferEval (alpine.js:118)
    at new Component (alpine.js:1423)
    at Object.initializeComponent (alpine.js:1870)
    at alpine.js:1857
    at alpine.js:1835
    at Array.forEach (<anonymous>)
    at Object.discoverUninitializedComponents (alpine.js:1834)
    at alpine.js:1856
    at NodeList.forEach (<anonymous>)
```
Move the component stuff into index.php with out the loading by htmx and it works just fine. The problem appears to be that it can't initialize the function dropdown. Don't know how to fix this so the component can load as a standalone.

**Reply by ryangjchandler** [reply]
Script tags inserted into the DOM via `innerHTML` and other methods won't be evaluated, therefore Alpine won't be able to find the `dropdown` function in the global namespace.

**Reply by KevinBatdorf** [reply]
Is `dropdown` available if you type it into the browser console? If yes, then it's probably loading too late, so it's not there when Alpine is looking for it.  

edit: Just saw Ryan's reply above mine, so if that's the case you just need to load things differently.

**Reply by Saeven** [reply]
@jsantari you solve this one?  I'm running into a similar issue.  Interestingly, loading partials like this with jQuery.load works just fine.

**Reply by jsantari** [reply]
No never did - I was just trying some things out so it wasn't a priority and I didn't bother to pursue it further.

**Reply by dmzone** [reply]
window.dropdown =... maybe?

**Reply by BartChris** [reply]
Maybe the latest changes in HTMX allow that? bigskysoftware/htmx#603
